conn.go

```diff
@@ -76,24 +79,113 @@ func (r *filteredConn) ReadFrom(b []byte) (n int, addr net.Addr, err error) {
 	select {
 	case <-timeout:
 		return 0, nil, errTimeout
-	case pkt := <-r.recvBuffer:
-		n := pkt.n
-		err := pkt.err
-		if l := len(b); l < n {
-			n = l
-			if err == nil {
-				err = io.ErrShortBuffer
-			}
+	case msg := <-r.recvBuffer:
+		n, _, err := copyBuffers(msg, b, nil)
+
+		r.source.returnBuffers(msg.Message)
+
+		return n, msg.Addr, err
+	case <-r.closed:
+		return 0, nil, errClosed
+	}
+}
+
+func (r *filteredConn) ReadBatch(ms []ipv4.Message, flags int) (int, error) {
+	if flags != 0 {
+		return 0, errNotSupported
+	}
+
+	if len(ms) == 0 {
+		return 0, nil
+	}
+
+	var timeout <-chan time.Time
+
+	if deadline, ok := r.deadline.Load().(time.Time); ok && !deadline.IsZero() {
+		timer := time.NewTimer(deadline.Sub(time.Now()))
+		timeout = timer.C
+		defer timer.Stop()
+	}
+
+	var msgs []messageWithError
+
+	defer func() {
+		for _, msg := range msgs {
+			r.source.returnBuffers(msg.Message)
 		}
-		copy(b, pkt.buf[:n])
-		r.source.bufPool.Put(pkt.buf[:r.source.packetSize])
-		if pkt.oobBuf != nil {
-			r.source.bufPool.Put(pkt.oobBuf[:r.source.packetSize])
+	}()
+
+	// We must read at least one message.
+	select {
+	//goland:noinspection GoNilness
+	case <-timeout:
+		return 0, errTimeout
+	case msg := <-r.recvBuffer:
+		msgs = append(msgs, msg)
+		if msg.Err != nil {
+			return 0, msg.Err
 		}
-		return n, pkt.addr, err
 	case <-r.closed:
-		return 0, nil, errClosed
+		return 0, errClosed
 	}
+
+	// After that, it's best effort. If there are messages, we read them.
+	// If not, we break out and return what we got.
+loop:
+	for len(msgs) != len(ms) {
+		select {
+		case msg := <-r.recvBuffer:
+			msgs = append(msgs, msg)
+			if msg.Err != nil {
+				return 0, msg.Err
+			}
+		case <-r.closed:
+			return 0, errClosed
+		default:
+			break loop
+		}
+	}
+
+	for i, msg := range msgs {
+		if len(ms[i].Buffers) != 1 {
+			return 0, errNotSupported
+		}
+
+		n, nn, err := copyBuffers(msg, ms[i].Buffers[0], ms[i].OOB)
+		if err != nil {
+			return 0, err
+		}
+
+		ms[i].N = n
+		ms[i].N = nn
+		ms[i].Flags = msg.Flags
+		ms[i].Addr = msg.Addr
+	}
+
+	return len(msgs), nil
+}
+
+func copyBuffers(msg messageWithError, buf, oobBuf []byte) (n, nn int, err error) {
+	if msg.Err != nil {
+		return 0, 0, msg.Err
+	}
+
+	if len(buf) < msg.N {
+		return 0, 0, io.ErrShortBuffer
+	}
+
+	copy(buf, msg.Buffers[0][:msg.N])
+
+	// Truncate, probably?
+	oobn := msg.NN
+	if oobl := len(oobBuf); oobl < oobn {
+		oobn = oobl
+	}
+	if oobn > 0 {
+		copy(oobBuf, msg.OOB[:oobn])
+	}
```

Comment:
Why no error here if the provided oob buffer is too small?

Reply:
This is just a random decision, with no good merit or particular reason.